### PR TITLE
Support human readable datetime type for PinotDB

### PR DESCRIPTION
**Pull request by xiangfu0**
### CATEGORY

Choose one

- [ ] Bug Fix
- [x] Enhancement (new features, refinement)
- [ ] Refactor
- [ ] Add tests
- [ ] Build / Development Environment
- [ ] Documentation

### SUMMARY
Support human readable datetime type for PinotDB
Now we can support datetime like '%Y-%m-%d' for time Series.
image

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Changes UI
- [ ] Requires DB Migration.
- [ ] Confirm DB Migration upgrade and downgrade tested.
- [ ] Introduces new feature or API
- [ ] Removes existing feature or API

### CODE REVIEW

File: superset/db_engine_specs/pinot.py [hunk 1]
```diff
@@ -44,13 +45,32 @@ def get_timestamp_expr(
...
-        seconds_or_ms = "MILLISECONDS" if pdf == "epoch_ms" else "SECONDS"
-        tf = f"1:{seconds_or_ms}:EPOCH"
+        tf = ""
+        if not is_epoch:
+            try:
+                today = datetime.datetime.today()
+                today.strftime(str(pdf))
+            except ValueError:
+                raise NotImplementedError(
+                    "Pinot currently doesn't support date formate: " + str(pdf)
```
Comment (villebro) [inline superset/db_engine_specs/pinot.py]: Typo (formate -> format) + please use f-string. Also, the error message indicates that the error is specific to Pinot; it is my understanding that `ValueError` in this case will be thrown if it is not a valid python date format.

**Review by villebro — CHANGES_REQUESTED** [review]
A few first pass comments

File: superset/db_engine_specs/pinot.py (hunk 1 above)
Comment (fx19880617) [reply superset/db_engine_specs/pinot.py]: You are correct, now we support datetime format, so will raise a ValueError here.

File: superset/db_engine_specs/pinot.py [hunk 2]
```diff
@@ -44,13 +45,32 @@ def get_timestamp_expr(
...
+                )
+            java_pdf = (
+                str(pdf)
+                .replace("%Y", "yyyy")
+                .replace("%m", "MM")
+                .replace("%d", "dd")
+                .replace("%H", "HH")
+                .replace("%M", "mm")
+                .replace("%S", "ss")
+            )
```
[the remark this answers is not in the document]
Comment (fx19880617) [reply superset/db_engine_specs/pinot.py]: done

File: superset/db_engine_specs/pinot.py [hunk 3]
```diff
@@ -39,18 +40,41 @@ class PinotEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
...
-        seconds_or_ms = "MILLISECONDS" if pdf == "epoch_ms" else "SECONDS"
-        tf = f"1:{seconds_or_ms}:EPOCH"
+        tf = ""
+        if not is_epoch:
+            try:
+                today = datetime.datetime.today()
+                today.strftime(str(pdf))
+            except ValueError:
+                raise ValueError(f"Invalid column datetime format:{str(pdf)}")
+            java_pdf = str(pdf)
```
Comment (villebro) [inline superset/db_engine_specs/pinot.py]: very nitty: the p should not be here, perhaps call write full name, i.e. `java_date_format` or similar

File: superset/db_engine_specs/pinot.py [hunk 4]
```diff
@@ -39,18 +40,41 @@ class PinotEngineSpec(BaseEngineSpec):  # pylint: disable=abstract-method
         "P1Y": "1:YEARS",
     }
 
+    _python_to_java_time_patterns = {
```
Comment (etr2460) [inline superset/db_engine_specs/pinot.py]: Let's type this as a `Dict[str, str]`?

**Review by villebro — APPROVED** [review]

**Pull request merged**